js/src/path-test.js

```diff
@@ -8,100 +8,277 @@ import {assert} from 'chai';
-    await assertPathEqual(false, v2, new Path().addField('v1').addField('bar'));
-    await assertPathEqual(203, v2, new Path().addField('v1').addField('baz'));
-    await assertPathEqual(undefined, v2, new Path().addField('v1').addField('notHere'));
-    await assertPathEqual(undefined, v2, new Path().addField('notHere').addField('foo'));
+    await assertResolvesTo('foo', v2, '.v1.foo');
+    await assertResolvesTo(false, v2, '.v1.bar');
+    await assertResolvesTo(203, v2, '.v1.baz');
+    await assertResolvesTo(undefined, v2, '.v1.notHere');
+    await assertResolvesTo(undefined, v2, '.notHere.foo');
   });
 
-  test('list', async () => {
-    const v = new List([1, 3, 'foo', false]);
+  test('index', async () => {
+    let v: Value;
+    const resolvesTo = async (exp: ?Value, val: Value, str: string) => {
+      // Indices resolve to |exp|.
+      await assertResolvesTo(exp, v, str);
+      // Keys resolves to themselves.
+      if (exp != null) {
+        exp = val;
+      }
+      await assertResolvesTo(exp, v, str + '@key');
+    };
+
+    v = new List([1, 3, 'foo', false]);
+
+    await resolvesTo(1, 0, '[0]');
+    await resolvesTo(3, 1, '[1]');
+    await resolvesTo('foo', 2, '[2]');
+    await resolvesTo(false, 3, '[3]');
+    await resolvesTo(null, 4, '[4]');
+    await resolvesTo(null, -4, '[-4]');
 
-    await assertPathEqual(1, v, new Path().addIndex(0));
-    await assertPathEqual(3, v, new Path().addIndex(1));
-    await assertPathEqual('foo', v, new Path().addIndex(2));
-    await assertPathEqual(false, v, new Path().addIndex(3));
-    await assertPathEqual(undefined, v, new Path().addIndex(4));
-    await assertPathEqual(undefined, v, new Path().addIndex(-4));
+    v = new Map([
+      [1, 'foo'],
+      ['two', 'bar'],
+      [false, 23],
+      [2.3, 4.5],
+    ]);
+
+    await resolvesTo('foo', 1, '[1]');
+    await resolvesTo('bar', 'two', '["two"]');
+    await resolvesTo(23, false, '[false]');
+    await resolvesTo(4.5, 2.3, '[2.3]');
+    await resolvesTo(null, 4, '[4]');
   });
 
-  test('map', async () => {
-    const v = new Map([[1, 'foo'], ['two', 'bar'], [false, 23], [2.3, 4.5]]);
+  test('hash index', async () => {
+    const b = true;
+    const br = new Ref(b);
+    const i = 0;
+    const str = 'foo';
+    const l = new List([b, i, str]);
+    const lr = new Ref(l);
+    const m = new Map([
+      [b, br],
+      [br, i],
+      [i, str],
+      [l, lr],
+      [lr, b],
+    ]);
+    const s = new Set([b, br, i, str, l, lr]);
+
+    const resolvesTo = async (col: Value, exp: ?Value, val: Value) => {
+      // Values resolve to |exp|.
+      await assertResolvesTo(exp, col, hashIdx(val));
+      // Keys resolves to themselves.
+      if (exp != null) {
+        exp = val;
+      }
+      await assertResolvesTo(exp, col, hashIdx(val) + '@key');
+    };
+
+    // Primitives are only addressable by their values.
+    await resolvesTo(m, null, b);
```

Comment:
The correct expectation is `null` because I want `resolve` to always return null if it's not found, not undefined.